tools/rosbag/src/recorder.cpp

```diff
@@ -285,6 +299,18 @@ std::string Recorder::timeToStr(T ros_t)
 void Recorder::doQueue(const ros::MessageEvent<topic_tools::ShapeShifter const>& msg_event, string const& topic, shared_ptr<ros::Subscriber> subscriber, shared_ptr<int> count) {
     //void Recorder::doQueue(topic_tools::ShapeShifter::ConstPtr msg, string const& topic, shared_ptr<ros::Subscriber> subscriber, shared_ptr<int> count) {
     Time rectime = Time::now();
+
+    if(options_.custom_freq && custom_record_freq_.find(subscriber->getTopic())!=custom_record_freq_.end()){
+        if(topic_time_catcher_.find(subscriber->getTopic())==topic_time_catcher_.end()){
+            topic_time_catcher_.insert_or_assign(subscriber->getTopic(), ros::Time::now());
+        }
+        else if((ros::Time::now().toSec() - topic_time_catcher_.at(subscriber->getTopic()).toSec())> (1.0f/custom_record_freq_.at(subscriber->getTopic()))){
+            topic_time_catcher_.insert_or_assign(subscriber->getTopic(), ros::Time::now());
```

Comment:
- `insert_or_assign` doesn't make too much sense; after above check, you already know the entry exists (maybe save the iterator to the element and reuse it
- I think it would make more sense to instead of saving the current time, save the next valid time:
``` 
else if(rectime > topic_time_catcher_.at(subscriber->getTopic())){
￼    topic_time_catcher_.insert_or_assign(subscriber->getTopic(), rectime + ros::Duration(1.0f/custom_record_freq_.at(subscriber->getTopic())));
}
```
- also seems to make more sense to store the period in the map instead of the frequency
- why not make a map with topic as key that contains both period and next_publish_time? Then you can remove the multiple lookups
- this is minor, but the way it is done now is slightly misleading; topics will be recorded at a lower frequency than specified, because the way it is actually done is enforcing a minimal time separation of messages

Reply:
I added the interval correction, which improved the performance of the recording.